internal/controller/infrastructure/remote_machine_controller.go

```diff
@@ -207,6 +235,82 @@ func (r *RemoteMachineController) Reconcile(ctx context.Context, req ctrl.Reques
 	return ctrl.Result{}, nil
 }
 
+func (r *RemoteMachineController) reservePooledMachine(ctx context.Context, rm *infrastructure.RemoteMachine) error {
+	pooledMachineList := &infrastructure.PooledRemoteMachineList{}
+	if err := r.Client.List(ctx, pooledMachineList, client.InNamespace(rm.Namespace)); err != nil {
+		return fmt.Errorf("failed to list pooled machines: %w", err)
+	}
+
+	var (
+		firstFreePooledMachine *infrastructure.PooledRemoteMachine
+		foundPooledMachine     *infrastructure.PooledRemoteMachine
+	)
+	for _, pm := range pooledMachineList.Items {
+		if pm.Spec.Pool == rm.Spec.Pool {
+			if pm.Status.Reserved && pm.Status.MachineRef.Name == rm.GetName() {
+				foundPooledMachine = &pm
+				break
+			}
+
+			if !pm.Status.Reserved {
+				firstFreePooledMachine = &pm
+			}
+		}
+	}
+
+	if foundPooledMachine == nil && firstFreePooledMachine == nil {
+		return ErrPooledMachineNotFound
+	}
+
+	if foundPooledMachine == nil && firstFreePooledMachine != nil {
+		foundPooledMachine = firstFreePooledMachine
+		foundPooledMachine.Status.Reserved = true
+		foundPooledMachine.Status.MachineRef = infrastructure.RemoteMachineRef{
+			Name:      rm.GetName(),
+			Namespace: rm.GetNamespace(),
+		}
+
+		err := r.Status().Update(ctx, foundPooledMachine)
+		if err != nil {
+			return fmt.Errorf("failed to update pooled machine status: %w", err)
+		}
+	}
+
+	rm.Spec.Address = foundPooledMachine.Spec.Machine.Address
+	rm.Spec.Port = foundPooledMachine.Spec.Machine.Port
+	rm.Spec.User = foundPooledMachine.Spec.Machine.User
+	rm.Spec.SSHKeyRef = foundPooledMachine.Spec.Machine.SSHKeyRef
+
+	return nil
+}
+
+func (r *RemoteMachineController) returnMachineToPool(ctx context.Context, rm *infrastructure.RemoteMachine) error {
+	pool := rm.Spec.Pool
```

Comment:
Should we bail out early if the RemoteMachine has no pool defined?

Reply:
Done